### ISSUE TITLE
Missing locale items - improve locale/translation system?

### ISSUE THREAD

**Issue by AndyMender** [Tuxemon/Tuxemon#462]
Upon starting the game with default settings I get the following error:
```
2018-10-06 21:19:52,989 - tuxemon.core.components.locale - ERROR - Key 'npc_red_name' does not exist in 'en_US' locale file.
```

Related locale issue: #409

I noticed the locale files are actually JSON files and contain a lot more than just regular translations (for instance, dialog options). It might be worthwhile to revisit the locale system and implement something more robust. A former project I participated in, Unknown Horizons, used Weblate and .po files to contain the translated text. Translation was carried out through simple text wrappers. I could try implementing something similar in Tuxemon.

Thoughts? :)

**EDIT:** I did some research and established a basic in-code translation framework using PO, POT and MO files compatible and based of `gettext`. Since some of the in-game text requires inserting variables into templates, I believe using `Jinja2` would be a good fit here :). All that's left is inserting the text wrappers around in-game text and generating text templates.

**Comment by bitcraft** [comment]
The current translation system was put together in a day or two, just to get the feature working.  We have considered Jinja2 in the past.  There are a few threads in the issues (hopefully not closed), that detail the limitations.  I'll list off a couple from memory:
* cannot handle gendered nouns
* capitalization rules
* pluralization for nouns, etc.

I'm not super familiar with gettext, but as I recall it is basically used everywhere and would be a good fit.  If you'd like to work with the translation system, feel free to.

**Comment by AndyMender** [comment]
Thank you for the feedback :). I'm actually already on it!

**Comment by AndyMender** [comment]
As a side note, I added a `constants` module to the `tuxemon` top module to handle directory/file paths, leveraging `os.path`. Regular string concatenation is extremely risky. All standardized and reusable paths are now accessible via the `tuxemon.constants.paths` module.

**Comment by AndyMender** [comment]
So apparently it's not so easy to create POT translation templates using some of Python's default tools, because the text is often hidden in JSON database files and is not readily accessible in the code. That's why it takes so long... 

The other problem is that the current translation system extracts information from the locale JSON and from the JSONs for specific objects (items, monsters, NPCs, etc.). With `gettext` based translations this is no longer necessary. The locale PO and MO files contain a full list of all translated strings (avoids duplication if a string is called multiple times in the code) for a particular locale and the JSON files defining objects can then store the actual details, not code keys like in the below JSON:
```json
{
  "slug": "item_apple",
  "name_trans": "item_apple_name",
  "description_trans": "item_apple_descr",
  "execute_trans": "combat_used_x",
  "success_trans": "generic_thing_success",
  "failure_trans": "generic_thing_failure",
  "effects": [
    "heal"
  ],
  "power": 20,
  "sort": "food",
  "sprite": "gfx/items/apple.png",
  "target": {
    "enemy monster": 1,
    "enemy team": 0,
    "enemy trainer": 0,
    "own monster": 2,
    "own team": 0,
    "own trainer": 0,
    "item": 0
  },
  "type": "Consumable",
  "usable_in": [
    "MainCombatMenuState",
    "WorldState"
  ]
}
```

In addition, templated strings like "${{name}} took ${{status}} damage!" can be put into the game code. They will receive a `gettext` translation tag, which will link them to a string in a translation file and template filling can be executed either via `Jinja2` or regular `.format()` calls.

**Comment by bitcraft** [comment]
Would you mind demonstrating what that would look like?  I'd also prefer to not have any localized strings in the code, english or otherwise.  Using slugs as it now allows non-programmers to edit even default game messages without digging into the code and Id like to ensure that doesn't change.

**Comment by bitcraft** [comment]
I am a outsider to the translation world, so I realize that there are conventions in other places.  So far the slugs have been working great and I don't see a reason right now to remove them.  From what you've presented, I don't see any advantages over what is currently used.

Its been a priority to remove all in-game strings and make all parts of the game localized without editing source code.  If we move the english locale back into the source code and need to change the English translation, we would have to edit the source code and the PO files.  I am pretty firm in keeping strings out of the code, and I think slugs reduce the maintenance overall.

If you are suggesting moving the translated format strings to the JSON files where they are used (put the strings for "potions" in the potion.json), then I can see the value in that:
```
{
  "slug": "Potion",
  "description_trans": "item_potion_descr",
  "execute_trans": "combat_used_x",
  "success_trans": "generic_thing_success",
  "failure_trans": "generic_thing_failure",
  "locale": {
    "en_US": {
      "description": "Heals a monster by 50 HP."
    }
  }
  "effects": [
    "heal"
  ],
  "power": 50,
  "sort": "potion",
  "sprite": "gfx/items/potion.png",
  "target": {
    "enemy monster": 1,
    "enemy team": 0,
    "enemy trainer": 0,
    "own monster": 2,
    "own team": 0,
    "own trainer": 0,
    "item": 0
  },
  "type": "Consumable",
  "usable_in": [
    "MainCombatMenuState",
    "WorldState"
  ]
}
```

For now, I think just fixing the translation "engine" to support automatic plural/gendered nouns, etc would be a bigger win for the game.

Just a related though, I think most the the "_trans" suffixes on the slugs is mostly legacy now.  At one time, slugs and raw unlocalized strings were mixed together and the "_trans" suffix/tag was useful for keeping the separate.   A I will open a followup issue to clean up the slug names so they are shorter.  For instance, slug names could be determined by a simple formula, like `"thing slug name" +"_"+ "descr"` would be the things translated description:  `"potion_descr"`.  This could be applied to all parts of the JSON DB and reduce the number of explicit tags/strings required in the the JSON entries.

**Comment by AndyMender** [comment]
> It's been a priority to remove all in-game strings and make all parts of the game localized without editing source code. If we move the english locale back into the source code and need to change the English translation in the future, we would have to edit the source code and the PO files. Seems like a more effort compared with slugs, which are unique and don't need to be changed once set. I am pretty firm in keeping strings out of the code, and I think slugs reduce the maintenance overall.

I accept this and fully embrace the need for easier code maintenance. The `slug` system would stay, but would be simplified. Currently, the translations are pulled from locale JSON files and matched via hackish field name concatenation. As some of the issues highlighted, this requires that every single locale JSON is adjusted upon in-code change and that alterations to JSON fields like 'combat_used_x' need to be propagated across multiple JSON files. One of the main advantages of using PO files and POT template files is that 1 POT file can be propagated to multiple PO files very easily. In addition, the issue of missing translation keys is virtually non-existent, because at least an empty string is provided.

The way I see it at the moment is that each item/monster/NPC could have a single JSON file with its full text description, etc. and templated dialogs would be stored in its own `dialogs.json` file.

I understand that this is quite challenging, nevertheless I would give it a try :). Worst case scenario, it provides easier means for building new translations. I will try to deliver sufficient documentation so that other contributors can decide about the usefulness of the additions.

**Comment by AndyMender** [comment]
I created a PR and shortly demonstrated the use of the new translation system with the `Item` class. if something is still not clear, please let me know and I will attempt to clarify :).

**Comment by bitcraft** [comment]
Thank you for opening the WIP PR.  Looking it over, it looks like you've dropped the slugs and instead are using english strings all over?  What I am worried about is whenever the English translation is changed, all locales, game JSON, etc will have to be changed as well.  Can you confirm that it will not be a hassle to change the english locale?  Remember also, that our map files also use slugs, and and dialog changes in the map files for English would require changes all over the game.

**Comment by bitcraft** [comment]
To clarify, please demonstrate how this fixes our issues with translations:
* Unable to define plurals automatically and correctly
* Unable to correctly handle gendered nouns

**Comment by bitcraft** [comment]
> this requires that every single locale JSON is adjusted upon in-code change and that alterations to JSON fields like 'combat_used_x' need to be propagated across multiple JSON files

This is no different than changing any typos or wording in English strings.  You would still have to change all the PO files if a source string changed.  Translations are subject to much more churn than slugs.  Slugs don't change.  I don't think I have ever changed a slug once it was created, but I have seen the English wording changed often.

> In addition, the issue of missing translation keys is virtually non-existent, because at least an empty string is provided.

I don't see how this fixes the issue with missing translations.  If it is never created, for either system, there is still an error.  Currently missing keys returns the slug, which is arguably better than an empty string.

> matched via hackish field name concatenation

I don't agree that it is hackish.  I would agree it is overly verbose, but it is easy to understand and doesn't require fluency in English to use.

Since this game is an open source project by volunteers, I want to eliminate all time wasting parts. 
* Suppose a map creator makes a typo in the map
* Nobody notices it
* Translator use the typo in their translations
* Map creator fixes typo
* Now all translations are broken

Using slugs, which never change once created, their is no risk to fixing dialogs, game interface strings, etc.  There are some people here who only do translations, and I don't want them to constantly tweak their files because the English text is changing.

Same goes for the map creators.  If they are making maps, suddenly they have more maintenance because they fix their maps, then fix the PO files, then run some script to propagate changes, etc.

We simply do not have the man power or time to build a cathedral system of gettext as you are proposing it, in my opinion.

Beside the maintenance burden of using English as `msgids` it also discourages people from contributing if they are not fluent English speakers.  I am also bilingual, and I want to make sure everyone gets an opportunity to contribute, even if their English is not great.

Since we are a multilingual project, suppose somebody want to make a map.  Now they are expected to be fluent in English to create dialogs.  Using slugs, they could create a translation in their own language, and also a broken English translation that is easily fixed later.  Using gettext in your proposal means you must be fluent in English to contribute anything since English strings are the key, and I'm not really comfortable with that.

**Comment by AndyMender** [comment]
I believe both approaches have their shortcomings, but considering your very specific expectations, I have to concede that the `slug` system is the more flexible one. The `slug`s could be made less verbose or more meaningful, but that's about it.

Would tidying up the `slug` system a bit be a meaningful effort to you?

**Comment by bitcraft** [comment]
Here is a list of problems with the slug, IMO.  This list could be split off into separate issues later:

* Overly verbose: many slugs are formulaic in creations, and we could add some magic to find them automatically.  For example, no need to specify "tux_name_trans" in the Tux JSON, just look for "tux_name_trans" in the PO
* "_trans" suffix is redundant.  should all be removed and all dialogs/strings expect slugs
* Not documented
* Because our dialogs don't support pagination, we have ended up with a lot of "dialog_1", "dialog_line_2", type slugs that should really be condensed into one.  With alternative fonts and translations, we cannot expect to have the same line length across the game.   So we need to rethink the dialog string markup.
* We have tools now to verify translations through the JSON, but we should also have a static analysis of the source code to check for missing slugs.

**Comment by AndyMender** [comment]
> Can gettext be used with slugs? Replace the English text phrase with the slug? I think using PO files will be easier for most people because of the restrictions of the JSON, so I don't really mind integrating gettext.

The new translator class I created called `TranslatorPo` allows wrapping any string source (also `slug`s, yes!) with a `translate` call and replacing the matching `msgid` string with the `msgstr` from a PO file. A new locale can be easily loaded with `load_locale`, but then the items, dialogs, monsters, NPC, etc. need to be reloaded so that the translation kicks in again. This is, of course the case for other translation systems as well.

This does not solve the gender and plural form issues, because those need to be marked and treated separately. It does make handling translations in general slightly easier, but that's about it I think :).

`msgstr` keys can span multiple lines, using the standard `\n` endline symbol. This could potentially take care of the dialog pagination issue.

Essentially, something like a mix of the `gettext` and `slug` based translation systems might work.

**Comment by bitcraft** [comment]
> msgstr keys can span multiple lines, using the standard \n endline symbol. This could potentially take care of the dialog pagination issue.

That would be great.  I think JSON is the wrong tool in the case of translations, and allowing line breaks could make mapping a lot easier.

> Essentially, something like a mix of the gettext and slug based translation systems might work.

I sort of envision slugs used everywhere like this:
 ```
# Super Potion item name
msgid "super_potion_name"
msgstr "Super Potion"

msgid "super_potion_description"
msgstr "Recovers 50 HP"
```

**Comment by AndyMender** [comment]
> I sort of envision slugs used everywhere like this:
```PO
# Super Potion item name
msgid "super_potion_name"
msgstr "Super Potion"

msgid "super_potion_description"
msgstr "Recovers 50 HP"
```

> The msgids would be formulaic. Attach base_slug + qualifier and let game engine find them dynamically. For example base_slug is "super_potion", qualifier is "name". Should be documented, but is easy to work with IMO.

I started implementing something similar already :). I used `super_potion` as the "name" qualifier, but this can be changed any time. With PO files we don't need to do any JSON key parsing anymore.

**Comment by bitcraft** [comment]
I can't think of any issues with using the slug as the name (no qualifier).

**Comment by AndyMender** [comment]
As I mentioned in the PR, the new translation system is pretty much finished. The only problem is that there are numerous discrepancies between slugs across existing locale JSON files. I used the en_US locale as reference since it seemed the most complete. That was not quite the case for the cs_CZ locale, though (the other one I finished).

**Comment by bitcraft** [comment]
Could you give me an example of a discrepancy?  I am aware that most translations are not 100%, and even the English one has issues, but you are correct, that is the most complete.

I'm going to check out the changes and merge it later this week.

**Comment by AndyMender** [comment]
> Could you give me an example of a discrepancy? I am aware that most translations are not 100%, and even the English one has issues, but you are correct, that is the most complete.

Sure! Here is the professor dialog from the en_US JSON:
```json
{
    "professor_dialog": [
        "Professor: \n Ah, yes. The very person I've been wanting to meet.",
        "I've heard that you want to learn about Tuxemon.",
        "Tuxemon are little animals, some call them monsters, that we befriend and use to battle other Tuxemon.",
        "I was wondering, I just received three new Tuxemon for research and I would like to give you one.",
        "Which one would you like?"
    ]
}
```
And here is the corresponding dialog from the cs_CZ JSON:
```json
{
    "professor_dialog1": "Profesor: \n Hurá. Přesně ten člověk se kterým jsem se chtěl setkat.",
    "professor_dialog2": "Slyšel jsem že se chceš učit o Tuxemonech.",
    "professor_dialog3": "Tuxemoni jsou malá zvířata, někteří je nazývají monstry, se kterými se přátelíme a používáme je k boji s dalšími Tuxemony."
}
```
The Tiled map file has an event for `professor_dialog`, but not for the professor dialogues listed in the cs_CZ locale JSON. There are many more cases like this one.

> I'm going to check out the changes and merge it later this week (after you've removed the WIP status, of course)

Would you like me to convert all other locales to the new translations system? I could do it over the weekend, I think.

**Comment by bitcraft** [comment]
Thanks for pointing that out.  I can see why you've been careful with translations.

The several professor dialogs were a hack from where there was a 1:1 limit between dialogs and strings.  So there needed to be 3 strings, one for each dialog.  Its not possible to supply a list of strings, and each will become a new dialog.  So you are looking at an unmaintained translation.

> Would you like me to convert all other locales to the new translation system? I could do it over the weekend, I think.

Do you mean JSON to PO?  yes, please use a script and include that in the scripts folder.  If you mean a general cleanup, its up to you.  Personally I would not update them manually.  If you would like to fix up a translation for your own mother tongue thats fine.

When the translation feature was first started, I made an announcement to warn people it is unstable, and will likely change, so any translations should be carried out at their own risk, meaning your work might not be used or will be obsolete.  Since the translations are independent of the game code and story, its up to the translators to keep it up to date.

The story itself is also subject to change, so I don't really encourage people to translate the maps until they have been vetted by the community.  Monster/Tuxemon descriptions, name, interface strings are more stable and those I'd say are ok to translate....but speech dialogs and map content changes too much, so i don't encourage it.

Anyway, its up to you, but if the you can use a script to just translate the json to PO, (please include with pr....script code is not really vetted too much) that would be good enough to have it all merged in, IMO.  Any work you do beyond that isn't really needed.

**Comment by AndyMender** [comment]
I'll see how feasible it is to create such a JSON-to-PO transition script, because some of the overly verbose slugs were simplified during refactoring to the new translation system. If it proves too time-consuming, I'll just leave it be.

**Comment by bitcraft** [comment]
Sounds good to me.

**Comment by AndyMender** [comment]
Since this is dragging on a bit and I didn't find any easy way to convert JSON content to PO templates, I'm removing the `WIP` flag from the PR. If there are any bugs, I will try to fix them :).

**Comment by bitcraft** [comment]
If im not mistaken, should it be a straightforward iteration over the key/value pairs and using print statements to write out the PO syntax?  I can't merge this unless the translations are at least valid.  If they are not pretty that is one thing, but if this PR leaves all translations totally broken (minor issues are OK), then i can wait until it is fixed.

**Comment by AndyMender** [comment]
Unfortunately, it's not as simple as dumping key/value pairs into the PO files, since the PR also reduces the verbosity of most slugs, including tuxemon, item, technique and NPC slugs. To create new translations, the en_US PO has to be used as a base.

Yes, only the en_US and cs_CZ locales work. The rest is currently broken (no PO files available, hence en_US is used as a fallback locale). I will try to create translations for the other locales as well.

**Comment by airon90** [comment]
I invited contributors to update their translations in #490. 

> Yes, only the en_US and cs_CZ locales work

What's the problem with Italian one? :)

**Comment by AndyMender** [comment]
The it_IT locale JSON is missing some keys present in the en_US locale JSON, but more importantly, it uses an older format for certain keys, like the first part of the professor dialog (1-5). Those keys were merged together and will remain merged in the PO templates. Unfortunately, it's a clean-up procedure I had to introduce, otherwise the affected locale files would remain invalid.

**Comment by airon90** [comment]
@AndyMender @bitcraft 
1. Is this issue open yet? 
2. Could you please explain what should I do to translate Tuxemon? I never used po files.

**Comment by bitcraft** [comment]
Most of the work has been done now.  Please open a new ticket for issues in the future.
